inc/hclib-forasync.h

```diff
@@ -62,539 +62,171 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
-	loop_domain_t loop0 = loop[0];
-	loop_domain_t loop1 = loop[1];
-	for(int i=loop0.low; i<loop0.high; i+=loop0.stride) {
-		for(int j=loop1.low; j<loop1.high; j+=loop1.stride) {
-			lambda(i, j);
-		}
-	}
-}
-
-template <typename T>
-inline void forasync3D_runner(const loop_domain_t loop[3], T lambda) {
-	loop_domain_t loop0 = loop[0];
-	loop_domain_t loop1 = loop[1];
-	loop_domain_t loop2 = loop[2];
-	for (int i = loop0.low; i < loop0.high; i += loop0.stride) {
-		for (int j = loop1.low; j < loop1.high; j += loop1.stride) {
-			for (int k = loop2.low; k < loop2.high; k += loop2.stride) {
-				lambda(i, j, k);
-			}
-		}
-	}
-}
-
-template <typename T>
-inline void forasync1D_recursive(loop_domain_t* loop, T lambda,
-        place_t *place, hclib_future_t **future_list) {
-	int low = loop->low, high = loop->high, stride = loop->stride, tile = loop->tile;
-	//split the range into two, spawn a new task for the first half and recurse on the rest
-	if ((high-low) > tile) {
-		int mid = (high+low)/2;
-		// upper-half
-		// delegate scheduling to the underlying runtime
-        auto lambda_wrapper = [=]() {
-            loop_domain_t ld = {mid, high, stride, tile};
-            forasync1D_recursive<T>(&ld, lambda, place, future_list);
-        };
-
-        hclib::async_await_at(lambda_wrapper, place, future_list);
-		// update lower-half
-		//continue to work on the half task
-		loop_domain_t ld = {low, mid, stride, tile};
-		forasync1D_recursive<T>(&ld, lambda, place, future_list);
-	} else {
-		//compute the tile
-		loop_domain_t ld = {low, high, stride, tile};
-		forasync1D_runner<T>(&ld, lambda);
-	}
-}
-
-template <typename T>
-inline void forasync2D_recursive(const loop_domain_t loop[2], T lambda,
-        place_t *place, hclib_future_t **future_list) {
-	loop_domain_t loop0 = loop[0];
-	int high0 = loop0.high;
-	int low0 = loop0.low;
-	int stride0 = loop0.stride;
-	int tile0 = loop0.tile;
-
-	loop_domain_t loop1 = loop[1];
-	int high1 = loop1.high;
-	int low1 = loop1.low;
-	int stride1 = loop1.stride;
-	int tile1 = loop1.tile;
-
-	//split the range into two, spawn a new task for the first half and recurse on the rest
-	loop_domain_t new_loop[2];
-	bool new_loop_initialized = false;
-
-	if((high0-low0) > tile0) {
-		int mid = (high0+low0)/2;
-		// upper-half
-		new_loop[0] = {mid, high0, stride0, tile0};
-		new_loop[1] = {low1, high1, stride1, tile1};
-		// update lower-half
-		high0 = mid;
-		new_loop_initialized = true;
-	} else if((high1-low1) > tile1) {
-		int mid = (high1+low1)/2;
-		// upper-half
-		new_loop[0] = {low0, high0, stride0, tile0};
-		new_loop[1] = {mid, high1, stride1, tile1};
-		// update lower-half
-		high1 = mid;
-		new_loop_initialized = true;
-	}
-	// recurse
-	if(new_loop_initialized) {
-		// delegate scheduling to the underlying runtime
-        auto lambda_wrapper = [=]() {
-            forasync2D_recursive<T>(new_loop, lambda, place, future_list);
-        };
+union ForAsyncFpUnion {
```

Comment:
This change is part of #19. I actually want to get rid of this union altogether. If we split the C interface  to use 3 different functions for 1D, 2D, and 3D (which is how we have it in C++) then that would let us get rid of this nasty hack for converting function pointers to/from `void*`.